# Lukema2022 lukiomatematiikkaa ohjelmoimalla

## Luku 2 ehtolauseet ja toisto

Luvun 2 tehtävissä tutkitaan toistoa laskemalla lukujonoihin liittyviä esimerkkejä. Ehtolauseita tutkitaan esimerkiksi tehtävässä 2.8, jossa sanallinen tehtävänanto.

## Ehtolauseet

Ohjelman täytyy joskus suorittaa komentoja vain, jos jokin ehto täyttyy. Silloin käytetään `if`-rakennetta.

In [1]:
# Testaa ohjelman tulostetta muuttamalla muuttujan x arvoa:
x = 10

# Tarkistetaan, onko muuttujan x arvo 10? Jos ehto toteutuu, niin suoritetaan sisennetty lohko
if x == 10:
    x = x + 2
    print("Muuttujan x arvo if-ehdon sisällä :", x)
# muuttujan x arvoksi asetaan 2-kertaa suurempi arvo
x = 2 * x
print("Muuttujan x arvo ohjelman lopussa:", x)

Muuttujan x arvo if-ehdon sisällä : 12
Muuttujan x arvo ohjelman lopussa: 24


Ylläolevan koodia koodia vastaa "Jos x on yhtä suuri kuin kymmenen, lisää siihen kaksi ja tulosta x. Lopuksi joka tapauksessa kerro x kahdella."

Pythonissa if-lauseen syntaksissa on sana `if`, sen jälkeen jokin ehto ja kaksoispiste. If-lauseen sisällä olevat komennot sisennetään neljällä välilyönnillä.

`If`-lausetta voi jatkaa `elif`- ja`else`-rakenteilla. Else ei ota mitään ehtoa, vaan sen jälkeen tuleva koodi suoritetaan, jos if:n jälkeistä ei suoriteta.

In [2]:
x = 10
if x > 10:
    print("x on suurempi kuin 10")
else:
    print("x on pienempi tai yhtä suuri kuin 10")

x on pienempi tai yhtä suuri kuin 10


**If-else**-rakenne on siis **joko-tai**-rakenne. Aina jompi kumpi suoritetaan.

`Elif` on lyhenne sanoista `else if`. Se toimii samalla if-else-rakenne, mutta sitä on helpompi ketjuttaa.


In [3]:
x = -1

if x > 10:
    print("x on suurempi kuin kymmenen")
else:
    if x < 0:
        print("x on negatiivinen")

x on negatiivinen


tarkoittaa samaa kuin

In [4]:
if x > 10:
    print("x on suurempi kuin kymmenen")
elif x < 0:
    print("x on negatiivinen")

x on negatiivinen


Esimerkki if-elif-else -ketjusta. Alla oleva ohjelma laskee pisteet-muuttujan arvosta yo-arvosanan.

In [5]:
pisteet = 80

if pisteet < 28:
    print("I")
elif pisteet < 40:
    print("A")
elif pisteet < 52:
    print("B")
elif pisteet < 64:
    print("C")
elif pisteet < 75:
    print("M")
elif pisteet < 90:
    print("E")
else:
    print("L")

E


Yhdessä if-rakenteessa on oltava yksi if, niin monta elif:iä kuin tarvitaan ja korkeintaan yksi else. Jos elseä käytetään, sen täytyy olla aina viimeisenä. **Esimerkiksi seuraava ohjelma ei ole sallittu:**

In [2]:
# Esimerkki virheellisestä rakenteesta
x = 0
if x == 0:
    print("x on 0")
else:
    print("x ei ole 0")
elif x < 0:
    print("x on negatiivinen")

SyntaxError: invalid syntax (3238114802.py, line 7)

Yllä olevan esimerkin oikea rakenne:

In [8]:
x = 1
if x == 0:
    print("x on 0")
elif x < 0:
    print("x on negatiivinen")
else:
    print("x ei ole 0")

x ei ole 0


 Tee ohjelma, joka kysyy käyttäjältä luvun ja kertoo onko luku pariton vai parillinen.

In [9]:
# Kirjoita ohjelma tähän
luku = 5
# jaollisuus tutkitaan % operaattorilla. Se kertoo jakojäännöksen
if luku % 2 == 0:
    print("Luku on parillinen")
else:
    print("Luku on pariton")


Luku on pariton


### Tehtävä 2.1
Toteuta lipunmyyntiohjelma, joka ottaa syötteenä iän ja tulostaa lipun hinnan. Hinnasto: 


| Ikä        | Hinta |
|------------|-------|
| Alle 7     | 0€    |
| 7-17       | 1.40€ |
| 18-69      | 2.80€ |
| 70 tai yli | 1.40€ |


In [3]:
# malliratkaisu
ikä = 71

if ikä < 7:
    print("Hinta on 0 euroa.")
elif ikä < 18:
    print("Hinta on 1.4 eur.")
elif ikä < 70:
    print("Hinta 2.80 eur")
else:
    print("Hinta on 1.40 eur")

Hinta on 1.40 eur


## Silmukat

Silmukoilla voi toistaa komentoja. Pythonissa on kahdenlaisia silmukoita. `For`-silmukat ovat hyviä listan läpikäymiseen, mutta `while`-silmukat ovat usein parempia. Niitä käytetään, jos toistojen määrä ei ole tiedossa etukäteen.

Silmukoiden jälkeen ohjelman suoritus jatkuu normaalisti rivi riviltä.

In [5]:
# tulostetaan kokonaisluvut väliltä 0..5 for-silmukan avulla
for luku in range(6):
    print(luku)

print("Nyt homma on valmis!")

0
1
2
3
4
5
Nyt homma on valmis!


In [6]:
# tuodaan math-kirjaston funktiot käyttöön
import math

summa = 0

for i in range(6):
    # sqrt on neliöjuuri
    print("Neliöjuuri luvusta", i, "on", round(math.sqrt(i),2))

Neliöjuuri luvusta 0 on 0.0
Neliöjuuri luvusta 1 on 1.0
Neliöjuuri luvusta 2 on 1.41
Neliöjuuri luvusta 3 on 1.73
Neliöjuuri luvusta 4 on 2.0
Neliöjuuri luvusta 5 on 2.24


Ylläolevassa esimerkissä `print(luku)` ajettiin 6 kertaa.

`While`-silmukka toimii melkein kuin if. Esimerkki:

In [9]:
print("Tulostetaan luvun 2 potensseja, jotka ovat pienempiä kuin luku 100: ")
luku = 1

while luku < 100:
    print(luku)
    luku = luku * 2 # tässä lasketaan muuttujan luku uusi arvo

Tulostetaan luvun 2 potensseja, jotka ovat pienempiä kuin luku 100: 
1
2
4
8
16
32
64


`While`-silmukassa runkoa eli silmukan sisällä olevaa koodia ajetaan niin kauan kun **ehto on tosi**.

## Tehtäviä:

### Tehtävä 2.2
Aritmeettinen lukujono alkaa 5, 8, 11... 

a) Luettele lukujonon 30 ensimmäistä jäsentä.

In [11]:
a = 5
d = 3

for luku in range(1,31):
    #for lohko alkaa
    print(luku,".luku on",a)
    a = a + d

1 .luku on 5
2 .luku on 8
3 .luku on 11
4 .luku on 14
5 .luku on 17
6 .luku on 20
7 .luku on 23
8 .luku on 26
9 .luku on 29
10 .luku on 32
11 .luku on 35
12 .luku on 38
13 .luku on 41
14 .luku on 44
15 .luku on 47
16 .luku on 50
17 .luku on 53
18 .luku on 56
19 .luku on 59
20 .luku on 62
21 .luku on 65
22 .luku on 68
23 .luku on 71
24 .luku on 74
25 .luku on 77
26 .luku on 80
27 .luku on 83
28 .luku on 86
29 .luku on 89
30 .luku on 92


b) Mikä on sadan ensimmäisen jäsenen summa?  $S_n = \frac{a_1 + a_n}{2}\cdot n$



In [12]:
a = 5
d = 3
summa = 0

for luku in range(1,101):
    summa = summa + a
    a = a + d

print("Kaikkien lukujen suma on", summa)

Kaikkien lukujen suma on 15350


c) Kuinka moni jonon jäsenistä on pienempi kuin 500?

In [13]:
a = 5
d = 3

määrä = 0

while a < 500:
    #print(a)
    a = a + d
    määrä = määrä + 1

print("Lukujen lukumäärä on", määrä)

Lukujen lukumäärä on 165


### Tehtävä 2.4
Laske kaikkien nelinumeroisten seitsemällä jaollisten lukujen summa.

In [14]:
# 1.nelinumeroinen 7:lla jaollinen luku on 1001, koska 1001 / 7 = 143
# viimeinen 4-numeroinen luku on 9999
# kuvataan luvut lukujonona

a = 1001
d = 7

summa = 0

while a < 10000:
    summa = summa + a
    a = a + d

print("7:lla jaollisten 4-numeroisten lukujen summa:", summa)

7:lla jaollisten 4-numeroisten lukujen summa: 7071071


Tehtävän voi ratkaista myös seuraavasti

In [15]:
summa = 0

for luku in range(1000, 9999+1):
    #print(luku)
    if luku % 7 == 0:
        summa = summa + luku
        #print("bingo")


print("7:lla jaollisten 4-numeroisten lukujen summa on ",summa)

7:lla jaollisten 4-numeroisten lukujen summa on  7071071


### Tehtävä 2.5
Tutki, kuinka monta geometrisen lukujonon 3, 9, 27, 81, ... jäsentä on laskettava yhteen, että summan arvo ylittää 9999?

In [26]:
# Geometrinen lukujono
a = 3
#suhdeluku, jolla kerrotaan
q = 3

yläraja = 9999
summa = 0
määrä = 0

while summa < yläraja:
    print("luku on ", a)
    summa = summa + a
    print("summa on", summa)
    a = a*q
    määrä = määrä + 1 
    
print("Summan yläraja",yläraja ,"ylittyy, kun yhteenlaskettavien lukujen lukumäärä on ", määrä)
print("Lukujonon ", määrä, "yhteenlaskettavan jäsenen summa on ", summa,".")

luku on  3
summa on 3
luku on  9
summa on 12
luku on  27
summa on 39
luku on  81
summa on 120
luku on  243
summa on 363
luku on  729
summa on 1092
luku on  2187
summa on 3279
luku on  6561
summa on 9840
luku on  19683
summa on 29523
Summan yläraja 9999 ylittyy, kun yhteenlaskettavien lukujen lukumäärä on  9
Lukujonon  9 yhteenlaskettavan jäsenen summa on  29523 .


### Tehtävä 2.6
Ohjelma kysyy käyttäjältä suorakulmion leveyden ja korkeuden. Tämän jälkeen ohjelma kysyy haluatko laskea p=piirin  vai a=pinta-alan ? Tämän jälkeen ohjelma tulostaa tiedon näytölle.

### Tehtävä 2.7
Kunnan väkiluku kasvoi 1.2%ia vuodessa. Vuonna 2015 kunnassa oli 24500 asukasta. Tutki minä vuonna kunnan asukasluku ylittää 30 000 asukkaan rajan.

### Tehtävä 2.8
Kännykkäliittymän A kuukausimaksu on  4 euroa ja puhelumaksu on 0.09 euroa/min. Liittymässä B ei ole kuukausimaksua, mutta  puhelumaksu on 0.12 euroa/min.  Tee ohjelma, joka kysyy käyttäjältä kuukauden puheaikaa, ja ilmoittaa käyttäjälle kumpi liittymä A vai B kannattaa valita ja mitä kuukauden puhelut maksavat. 

### Tehtävä 2.9
Tarkastellaan lukujonoa $a_n =15n + 120, n=1, 2,3\dots$  Kirjoita ohjelma, joka vastaa seuraaviin kysymyksiin:

a) Ohjelma tulostaa jonon 10 ensimmäistä jäsentä.


b) Onko luku  12 455 lukujonon jäsen?


c)  Mistä jäsenestä alkaen lukujonon jäsenet ovat suurempia kuin  1000?

### Tehtäviä 2.10

Jaa luku 29282 alkutekijöihin

In [27]:
n = 29282
jakaja = 2

# Käydään läpi kaikki lukua n pienemmät luvut

while jakaja < n + 1:
    
    # Kokeillaan, jakaako jakaja luvun n
    
    if n % jakaja == 0:
        
        # Jos jakaja on alkutekijä, tulostetaan se ja aloitetaan läpikäynti alusta
        
        print(jakaja, "on alkutekijä")
        n = n // jakaja
        jakaja = 2
    else:
        
        # Jos jakaja ei ole alkutekijä, siirrytään seuraavaan lukuun
        
        jakaja = jakaja + 1

2 on alkutekijä
11 on alkutekijä
11 on alkutekijä
11 on alkutekijä
11 on alkutekijä


In [28]:
# tarkistus
2*11**4

29282

## Tehtävä 2.11

Vuosi on karkausvuosi, jos se on jaollinen 4:llä. Kuitenkin jos vuosi on jaollinen 100:lla, se on karkausvuosi vain silloin, kun se on myös jaollinen 400:lla.
Tee ohjelma, joka tulostaa kaikki karkausvuodet välillä 1800...2200, jokainen vuosi omalla rivillään.

Tuloste alkaa siis seuraavasti:

1808
1812
1816
1820
1824
1828
1832
1836
1840
1844
1848
1852
1856
1860
1864
1868
1872
1876

In [29]:
for vuosi in range(1800,2201):
    if vuosi % 4 == 0 and (vuosi % 100 != 0 or vuosi % 400 == 0):
        print(vuosi)

1804
1808
1812
1816
1820
1824
1828
1832
1836
1840
1844
1848
1852
1856
1860
1864
1868
1872
1876
1880
1884
1888
1892
1896
1904
1908
1912
1916
1920
1924
1928
1932
1936
1940
1944
1948
1952
1956
1960
1964
1968
1972
1976
1980
1984
1988
1992
1996
2000
2004
2008
2012
2016
2020
2024
2028
2032
2036
2040
2044
2048
2052
2056
2060
2064
2068
2072
2076
2080
2084
2088
2092
2096
2104
2108
2112
2116
2120
2124
2128
2132
2136
2140
2144
2148
2152
2156
2160
2164
2168
2172
2176
2180
2184
2188
2192
2196


## Tehtävä 2.12

Kalle on vuoden vanhempi kuin Jaakko, joka taas on vuoden vanhempi kuin Aapo. Heidän ikiensä tulo on 42840. Kuinka vanha Aapo on?

Vihje: Koska tiedetään, että ratkaisu on kokonaisluku, niin voidaan käydä läpi lukuja esim. väliltä 0..150, ja tutkia mitkä kokonaisluvuista toteuttavat ehdon.

In [30]:
# Oletetaan että Aapon ikä x on <150 vuotta.
for x in range(150):
    if x*(x+1)*(x+2) == 42840:
        print(x)

34


## Tehtävä 2.13

Etsi kokonaisluku x, joka ratkaisee yhtälön $x^3 - 60x^2 + 60x = 59$.


In [31]:
x = 0
löytyi = False

while not löytyi:
    if x**3 -60*x**2+60*x == 59:
        print(x)
        löytyi = True
    x = x +1

59


In [32]:
# ratkaisu 2
# etsitään ratkaisua väliltä 0..1999
for x in range(2000):
    if x**3-60*x**2+60*x == 59:
        print(x)

59


## Tehtävä 2.14

Mikä on pienin positiivinen kokonaisluku, joka on jaollinen luvuilla 12, 34 ja 56?

Silmukka voidaan keskeyttää `break`-komennolla, jos ratkaisu eli tutkittava ehto toteutuu. Ohjelman suoritus jatkuu silmukan jälkeen normaalisti.

In [33]:
# tehtävässä harjoitellaan silmukan keskeyttämistä break -komennolla
for x in range(1,20000):
    if x%12 == 0 and x%34 == 0 and x%56 == 0:
        print(x)
        break

print("Tehtävä on nyt ratkaistu")

2856
Tehtävä on nyt ratkaistu


### Tehtävä 2.15
Yo-tehtävä: Kahden positiivisen luvun $a$ ja $b$ geometrinen on $\sqrt{ab}$.

a) Anna esimerkki välin $2-100$ kahdesta luvusta, joille $\sqrt{ab}$ on kokonaisluku.

b) Satunnaislukugeneraattori arvoo toisistaan riippumatta kaksi lukua väliltä 1-100 niin, että jokaisen luvun todennäköisyys on $\frac{1}{100}$. Mikä on todennäköisyys, että arvottujen lukujen geometrinen keskiarvo on kokonaisluku? Voit esittää simulointiin perustuvan ratkaisun.